### PR TITLE
Write coverage data into .tox/coverage

### PR DESCRIPTION
**Pull request by wsanchez**
Write coverage data into `.tox/coverage` instead of writing lots of `.coverage.BLEARGHHUCKHUCKBARF` files in the top of the source directory.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,7 +2,6 @@
 /_trial_temp/
 /build/
 htmlcov
-.coverage
```
Comment (wsanchez) [inline .gitignore]: This should have been `.coverage.*` all along anyway…

File: .travis.yml [hunk 2]
```diff
@@ -33,7 +33,7 @@ matrix:
     - env: TOXENV=coverage-pypy-twtrunk,codecov   PYPY_VERSION=5.7.1
 
     - python: 2.7
-      env: TOXENV=coverage-py27-tw155
+      env: TOXENV=coverage-py27-tw155,codecov
```
Comment (wsanchez) [inline .travis.yml]: Whoops, this was missed in prior PR.

File: tox.ini [hunk 3]
```diff
@@ -56,14 +56,22 @@ setenv =
     PIP_DISABLE_PIP_VERSION_CHECK=1
     VIRTUALENV_NO_DOWNLOAD=1
 
+    coverage: COVERAGE_FILE={toxworkdir}/coverage/coverage.{envname}
+    codecov: COVERAGE_FILE={toxworkdir}/coverage/coverage
```
Comment (wsanchez) [inline tox.ini]: What going on here is we create a `coverage.<envname>` file for each environment.

When the `codecov` environment runs `coverage combine`, it'll combine any `coverage.*` files into a single `coverage` file.

**Comment by glyph** [comment]
What is the benefit of putting the coverage files in a nonstandard location? This seems like a step backwards from using -p, which allows for parallelism, whereas this implicitly disallows combination of coverage from multiple runs within the same environment.

**Comment by wsanchez** [comment]
@glyph: Every time you run tox right now, you add more `.coverage.*` files to the top of the source tree.  Tox results should be in `.tox`.  Also, the files currently are not distinguishable in terms of which environment generated which file.  This fixes both.

Anyway, `-p` doesn't actually parallelize anything, though.  We can do a directory per environment and use `-p` again, but there's not much point in that until we can parallelize tests within an environment.  Is that a thing you can do now?

**Comment by glyph** [comment]
> Every time you run tox right now, you add more .coverage.* files to the top of the source tree.

Right, and then you run `coverage combine` or `coverage erase` to clean them up.

> Tox results should be in .tox.

Should they?  The virtuale environments should be in `.tox` but other build artifacts routinely appear outside.  I don't like having to rummage around in various `.tox` directories before `coverage combine` will work as advertised.  (Although this personal opinion is something I don't give a lot of weight; if the standard way to do this were the way you're suggesting and I were the weird one here I'd gladly get rid of my customization in this case.)

> Also, the files currently are not distinguishable in terms of which environment generated which file. This fixes both.

This seems like more of an issue with `coverage` itself though; perhaps something in the filename should indicate something about an environment?  Separating them out in its current state inherently breaks `coverage combine` which seems like a bigger deal than leaving stale coverage files in place.

Perhaps you just want to add something to ~/.gitconfig core.excludesfile?

**Comment by wsanchez** [comment]
> Right, and then you run `coverage combine` or `coverage erase` to clean them up.

Better is for Tox to be set up to do that for you…  My point is that when you ask Tox to run an environment, it should give you accurate coverage results for that run.  The current setup has all of the environments in one pool and makes it hard to clean up.

> I don't like having to rummage around in various .tox directories before coverage combine will work as advertised.

Why would you want to do that instead of, say, running `tox -e coverage_combine`?  That isn't in this patch, because big patches don't get reviewed, but the intent is for all of this to be automated so you always get consistent results from Tox, not after you then also run some manual commands.

**Pull request closed without merging**